### PR TITLE
6213 Будюк В.А. Лаб.5 Вар.4

### PR DESCRIPTION
**Pull request by Nibuku**
Главное достижение данной лабы:

### CODE REVIEW

**Review by alxmcs — CHANGES_REQUESTED** [review]
Гораздо удобнее смотреть/анализировать результаты обучения модели, когда графики потерь для обучающей и валидационой выборок нарисованы в одних осях ( и для accuracy аналогично).

Так что для каждого эксперимента я бы соорудил что-то типа:
```py
import matplotlib.pyplot as plt
import numpy as np

x =  np.arange(1,101,1)
y1 = 1/(x+1)
y2 = 1/x 
y3 = 1 -1/(x+1) 
y4 = 1 -1/x

fig=plt.figure(figsize=(20, 5))
ax1 = fig.add_subplot(1,2,1)
plt.plot(x,y1, color='steelblue',label = 'train loss')
plt.plot(x,y2, color='orange',label = 'validation loss')
ax1.set_ylabel('MSE', size= 14)
ax1.set_xlabel('Epochs',size= 14)
plt.legend(loc=1, prop={'size': 14}) 
ax2 = fig.add_subplot(1,2,2)
plt.plot(x,y3, color='steelblue', label = 'train accuracy')
plt.plot(x,y4, color='orange', label = 'validation accuracy')
plt.legend(loc=4,prop={'size': 14}) 
ax2.set_ylabel('Accuracy',size= 14)
ax2.set_xlabel('Epochs',size= 14)
plt.tight_layout(pad=0.4, w_pad=0.5, h_pad=3.0)
```

image

**Review by alxmcs — CHANGES_REQUESTED** [review]
Осталось только проанализировать результаты:
1) По заданию надо что-то сказать насчет результатов экспериментов с размером батчей.
2) Есть предположения, почему accuracy на валидационной выборке больше, чем на обучающей?
image

**Comment by Nibuku** [comment]
> Осталось только проанализировать результаты:
> 
> 1. По заданию надо что-то сказать насчет результатов экспериментов с размером батчей.
> 2. Есть предположения, почему accuracy на валидационной выборке больше, чем на обучающей?
>    image

1. batch_size-это размер пакета данных, которые используются в одной эпохе обучения. Значит, чем больше размер батча-тем дольше обучается модель и тем выше её точность, но если размер батча будет слишком большой, то модель начнет переобучаться-запоминать обучающую выборку и в итоге на валидационной выборке ее точность значительно будет отличаться от обучающей. В моей лабе, если я правильно понимаю, лучше использовать batch_size 100 или 200, потому что при 300 точность валидационной выборки сильно падает по сравнению с обучающей, а потери растут. 
2. Вероятно, это связано с тем, что в обучающей выборке есть больше "мусора и шума" из-за ее размера (1608). А в валидационной мусора меньше => она оказалась проще для модели => точность выросла. 
(Ну и еще если модель начинает переобучаться, то она просто запоминает часть обучающей выборке, а так как в валидационной изображения +-такие же (т.к. оригинальных изображений роз тюльпанов как оказалось не так много), то она просто выдает ответ, который запомнила.)

**Review by alxmcs — APPROVED** [review]
Зачтено

Насчет 2), я бы еще предположил, что замешан [dropout](https://pytorch.org/docs/stable/generated/torch.nn.Dropout.html)
